### ISSUE TITLE
Improve routing system

### ISSUE THREAD

**Issue by jmjuanes** [opencb/jsorolla#283]
**Note**: this issue is still a WIP. API specification may change.

## Description (WIP)

At this moment we have a basic implementation of a routing system in IVA. 

There are some issues like #218 that will be easier to fix if we have an improved routing system.

## Proposed improvement

Implement a custom routing system based on web components. The idea is to have a web component that will be rendered only if the specified path template matches the current browsing URL. 

```html
<custom-route path="/home" exact render="${req => html`
    <home-page></home-page>
`}"></custom-route>
```

Routes inside routes are also supported:

```html
<!-- This route will match "/interpreter" and also "/interpreter/editor" -->
<custom-route path="/case-interpreter" render="${req => html`

    <!-- This route will only match "/interpreter/editor" -->
    <custom-route path="/case-interpreter/editor" exact render="${req => html`
        <case-interpreter-editor></case-interpreter-editor>
    `}"></custom-route>

    <!-- This route will only match "/interpreter" -->
    <custom-route path="/case-interpreter" exact render="${req => html`
        <case-interpreter-home></case-interpreter-home>
    `}"></custom-route>

`}"></custom-route>
```

To automatically force an update of the application when the URL changes, we should also implement a `custom-router` component. This component should be used **only once** in the application, and all routes must be inside this component:

```html
<custom-router>
    <!-- Routes should be placed here -->
    <custom-route path="..." render="${req => ...}"/></custom-route>
    <custom-route path="..." render="${req => ...}"/></custom-route>
    . . . 
</custom-router>
```

The `custom-router` component will be listening for `hashchange` events and will force an update when there are URL changes. See https://developer.mozilla.org/en-US/docs/Web/API/WindowEventHandlers/onhashchange.

### API specification

The `custom-route` component will accept the following properties:

- `path`: a `string` that describes the pathname that the route matches. This can also be a dynamic path string, see the **Dynamic routing section** for more info.
- `exact`: a `boolean` used to ensure that the route's path is an exact match of the current location. Read more at the **Exact routing matching section**.
- `render`: a function that will be called when the route matches.

### Render arguments

The `render` function will be called with a single argument `request`, which includes the information about the matched URL: 

- `params`: an object with the named parameters of the URL. See the **Dynamic routing** section.
- `query`: an object with the query-string values of the URL.

**Important note**: query strings are not part of the route path.

Example of a request object:

```
URL: /interpreter/case/interpretation1?mode=edit&showAll=true
TEMPLATE: /interpreter/case/:case

request = {
    params: {
        case: "interpretation1",
    },
    query: {
        mode: "edit",
        showAll: "true",
    },
}
```

### Dynamic routing

The `path` can also be a dynamic path string, using named parameters prefixed by a colon to the parameter name. For example, the path `/user/:name` matches `/user/bob` and `user/susan`. The captured values are stored in the `request.params` object passed as an argument to the `render` function.

Catch-all routes will be also supported:

```html
<custom-route path="*" render="${req => "Hello"}"/></custom-route>
``` 

### Exact routing matching

| Route path | Current path | Is Exact active? | Matches? |
|------------|--------------|------------------|----------|
| `/`        | `/one/two`   | Yes              | No       |
| `/`        | `/one/two`   | No               | Yes      | 
| `/one`     | `/one/two`   | Yes              | No       |
| `/one`     | `/one/two`   | No               | Yes      |
| `/one/two` | `/one/two`   | Yes              | Yes      |
| `/one/two` | `/one/two`   | No               | Yes      |

## Tasks

- [ ] Implement `custom-router` component. 
- [ ] Implement `custom-route` component.
- [ ] Implement routing utils.
- [ ] Replace IVA routing with `custom-route`.
